rtl/eclrtl/rtldynfield.cpp

```diff
@@ -1405,202 +1381,516 @@ class GeneralRecordTranslator : public CInterfaceOf<IDynamicTransform>
+{
+    GeneralRecordTranslator *ret = new GeneralRecordTranslator(destRecInfo, srcRecInfo);
+    ret->init(true);
+    return ret;
+}
+
+class CallbackRecordTranslator : public RecordTranslatorBase
+{
+public:
+    CallbackRecordTranslator(const RtlRecord &destRecInfo, const RtlRecord &srcRecInfo) : RecordTranslatorBase(destRecInfo, srcRecInfo)
+    {
+    }
+// IDynamicTransform impl.
+    virtual size32_t translate(ARowBuilder &builder, IVirtualFieldCallback & callback, const byte *sourceRec) const override
+    {
+        throwUnexpected();
+    }
+    virtual size32_t translate(ARowBuilder &builder, IVirtualFieldCallback & callback, const RtlRow &sourceRow) const override
+    {
+        throwUnexpected();
+    }
+    virtual size32_t translate(ARowBuilder &builder, IVirtualFieldCallback & callback, const IDynamicFieldValueFetcher & fetcher) const override
+    {
+        return doTranslate(builder, callback, 0, fetcher);
+    }
+    virtual bool canTranslate() const override
+    {
+        return (matchFlags & match_fail) == 0;
+    }
+    virtual bool needsTranslate() const override
+    {
+        return (matchFlags & ~(match_link|match_inifblock)) != 0;
+    }
+    virtual bool needsNonVirtualTranslate() const override
+    {
+        return (matchFlags & ~(match_link|match_virtual|match_keychange|match_inifblock)) != 0;
+    }
+    virtual bool keyedTranslated() const override
+    {
+        return (matchFlags & match_keychange) != 0;
+    }
+protected:
+    virtual RecordTranslatorBase *createRecordTranslator(const RtlRecord &destRecInfo, const RtlRecord &srcRecInfo) const
+    {
+        CallbackRecordTranslator *ret = new CallbackRecordTranslator(destRecInfo, srcRecInfo);
+        ret->init(false);
+        return ret;
+    }
+    static size32_t translateScalarFromUtf8(ARowBuilder &builder, size32_t offset, const RtlFieldInfo *field, const RtlTypeInfo &destType, const RtlTypeInfo &sourceType, const char *source, size_t srcLen)
+    {
+        switch(destType.getType())
+        {
+        case type_boolean:
+        case type_int:
+        case type_swapint:
+        case type_packedint:
+        case type_filepos:
+        case type_keyedint:
+        {
+            __int64 res = rtlUtf8ToInt(srcLen, source);
+            offset = destType.buildInt(builder, offset, field, res);
+            break;
+        }
+        case type_real:
+        {
+            double res = rtlStrToReal(srcLen, source);
+            offset = destType.buildReal(builder, offset, field, res);
+            break;
+        }
+        case type_data:
+        case type_string:
+            //Special case if source and destination types are identical to avoid cloning strings
+            if ((destType.fieldType & (RFTMkind|RFTMebcdic)) == sourceType.getType())
+            {
+                offset = destType.buildString(builder, offset, field, srcLen, source);
```

Comment:
All calls will need to be buildUtf8 (should possibly be called buildFromUtf8).  Also length is length in codepoints, rather than bytes, so need to call rtlUtf8Length(srcLen, source).  Parameter is clearer if it is called srcSize.

Reply:
right, will change.